### PR TITLE
feat: add support for application command permissions v2

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Credit to @vitorlops for making the basis for this PR.

**If applicable, please reference Discord API Docs PRs or commits that influenced this PR:**

- https://github.com/discord/discord-api-docs/pull/4830
- https://github.com/discord/discord-api-docs/pull/4836
- https://github.com/discord/discord-api-docs/pull/4833
- https://github.com/discord/discord-api-docs/pull/4845
- https://github.com/discord/discord-api-docs/pull/4861

### CODE REVIEW

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts
Comment (vladfrangu) [inline deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:61]:
```suggestion
		return String(guildId);
```

File: deno/payloads/v10/_interactions/applicationCommands.ts
Comment (vladfrangu) [inline deno/payloads/v10/_interactions/applicationCommands.ts:82]: Should deprecate this

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Comment by vladfrangu** [comment]
Should also track https://github.com/discord/discord-api-docs/pull/4836 and https://github.com/discord/discord-api-docs/pull/4833

File: payloads/v9/auditLog.ts
Comment (ckohen) [inline payloads/v9/auditLog.ts:525]: This one doesn't appear to actually exist, waiting for confirmation on it

File: payloads/v9/auditLog.ts
Comment (ckohen) [inline payloads/v9/auditLog.ts:531]: this one is an object currently, again waiting on confirmation

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts
Comment (vitorlops) [inline deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]:
```suggestion
	 * The id of the role, user or channel. Can also be a permission constant
```

File: deno/payloads/v10/_interactions/applicationCommands.ts
Comment (vitorlops) [inline deno/payloads/v10/_interactions/applicationCommands.ts:76]:
```suggestion
	default_member_permissions?: Permissions | null;
```
https://github.com/discord/discord-api-docs/pull/4830/commits/0dc91010bb37ed059a03a6ec52f6dc110ff3b684

File: deno/payloads/v10/oauth2.ts
Comment (vladfrangu) [reply deno/payloads/v10/oauth2.ts:115]: We usually copy paste doc strings directly from Discord so.. some strings may look out of place

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts
Comment (vladfrangu) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]: Is this documented on Discord's side? If not it should be

Comment (vitorlops) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:30]: It is yes
image

**Comment by vitorlops** [comment]
About https://github.com/discordjs/discord-api-types/pull/415#issuecomment-1111959991, more one to track https://github.com/discord/discord-api-docs/pull/4845

File: deno/payloads/v10/_interactions/applicationCommands.ts
Comment (ckohen) [reply deno/payloads/v10/_interactions/applicationCommands.ts:80]: This is changing, likely before this PR lands

File: rest/v10/interactions.ts
Comment (ckohen) [inline rest/v10/interactions.ts:76]: Did the wrong one here, should be all the guild endpoints - post, patch, and put

File: deno/payloads/v10/_interactions/_applicationCommands/permissions.ts
Comment (vladfrangu) [reply deno/payloads/v10/_interactions/_applicationCommands/permissions.ts:61]: This was not resolved

File: deno/payloads/v10/auditLog.ts
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:36]: Is this field always present (even for audit log entries that do not affect a command) or can it be undefined too?

File: deno/payloads/v10/auditLog.ts
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:537]: These two are wrong, one of the possibilities is now the `change.key` can be a Snowflake and have a value that for all intents and purposes we could only strictly type as unknown... (for this one PR the value is APICommandPermissionData or w/e)

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v10/auditLog.ts
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:532]: The key is a Snowflake. It's not 'snowflake'

File: deno/payloads/v9/_interactions/_applicationCommands/permissions.ts
Comment (vladfrangu) [inline deno/payloads/v9/_interactions/_applicationCommands/permissions.ts:63]:
```suggestion
	AllChannels: (guildId: string | bigint): Snowflake => String(BigInt(guildId) - 1n) 
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v10/auditLog.ts
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:36]: I was wrong, seems that this is always returned, only as an empty array 🥴

File: deno/payloads/v10/auditLog.ts
Comment (vladfrangu) [inline deno/payloads/v10/auditLog.ts:530]:
```suggestion
 * Represents a change where the key is a snowflake.
 * Currently, the only known instance of this is returned when permissions for a command were updated (<insert name of object here>)
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]
final requests i swear

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**
